plugins/pur/core/src/main/java/org/pentaho/di/repository/pur/PurRepository.java

```diff
@@ -3056,41 +3352,51 @@ protected void saveTransOrJob( ISharedObjectsTransformer objectTransformer, Repo
     RepositoryFile file;
     if ( isUpdate ) {
       ObjectId id = element.getObjectId();
-      file = pur.getFileById( id.getId() );
-      if ( checkLock && file.isLocked() && !unifiedRepositoryLockService.canUnlockFileById( id ) ) {
-        throw new KettleException( "File is currently locked by another user for editing" );
-      }
-      if ( checkDeleted && isInTrash( file ) ) {
-        // absolutely awful to have UI references in this class :(
-        throw new KettleException( "File is in the Trash. Use Save As." );
-      }
-      // update title and description
-      file =
-          new RepositoryFile.Builder( file ).title( RepositoryFile.DEFAULT_LOCALE, element.getName() ).createdDate(
-              versionDate != null ? versionDate.getTime() : new Date() ).description( RepositoryFile.DEFAULT_LOCALE,
-              Const.NVL( element.getDescription(), "" ) ).build();
+      readWriteLock.writeLock().lock();
+
       try {
+        file = pur.getFileById( id.getId() );
+
+        if ( checkLock && file.isLocked() && !unifiedRepositoryLockService.canUnlockFileById( id ) ) {
```

Comment:
Should really try to have the bail-out code before grabbing the write lock. So these two if statements and the get should be in a read lock try/finally block above. Then, you only grab the write lock when you need it.